basic_context.go

```diff
@@ -1,21 +1,83 @@
+// Copyright (C) 2022 Storj Labs, Inc.
+// See LICENSE for copying information.
+
 package clingy
 
 import (
+	"bytes"
 	"context"
 	"io"
+	"os"
+
+	"github.com/zeebo/errs/v2"
 )
 
-type basicContext struct {
+var stdinKey, stdoutKey, stderrKey struct{}
+
+// StdioContext can help to test CLI apps with wrapping stdout/stdin/stderr.
+type StdioContext struct {
 	context.Context
+}
+
+// StdioTestContext is for testing, where stdio is replaced with in-memory buffers.
+type StdioTestContext struct {
+	StdioContext
+}
 
-	stdin  io.Reader
-	stdout io.Writer
-	stderr io.Writer
+func WrapStdio(ctx context.Context) StdioContext {
+	ctx = context.WithValue(ctx, stdinKey, os.Stdin)
+	ctx = context.WithValue(ctx, stderrKey, os.Stderr)
+	ctx = context.WithValue(ctx, stdoutKey, os.Stdout)
+	return StdioContext{
+		Context: ctx,
+	}
 }
 
-func (b basicContext) Read(p []byte) (n int, err error)  { return b.stdin.Read(p) }
-func (b basicContext) Write(p []byte) (n int, err error) { return b.stdout.Write(p) }
+// WithBufferedStdio creates context with in-memory stdio targets.
+func WithBufferedStdio(ctx context.Context) StdioTestContext {
```

Comment:
This should not be part of the package, and is unnecessary because the Environment already allows this.